pump/storage/storage.go

```diff
@@ -344,22 +358,92 @@ func (a *Append) updateStatus() {
 	}
 }
 
+// Write a commit binlog myself if the status is committed,
+// otherwise we can just ignore it, we will not get the commit binlog while iterator the kv by ts
+func (a *Append) writeCBinlog(pbinlog *pb.Binlog, commitTS int64) error {
+	// write the commit binlog myself
+	cbinlog := new(pb.Binlog)
+	cbinlog.Tp = pb.BinlogType_Commit
+	cbinlog.StartTs = pbinlog.StartTs
+	cbinlog.CommitTs = commitTS
+
+	req := a.writeBinlog(cbinlog)
+	if req.err != nil {
+		return errors.Annotate(req.err, "writeBinlog failed")
+	}
+
+	// when writeBinlog return success, the pointer will be write to kv async,
+	// but we need to make sure it has been write to kv when we return true in the func, then we can get this commit binlog when
+	// we update maxCommitTS
+	// write the ts -> pointer to KV here to make sure it.
+	pointer, err := req.valuePointer.MarshalBinary()
+	if err != nil {
+		panic(err)
+	}
+
+	err = a.metadata.Put(encodeTSKey(req.ts()), pointer, nil)
+	if err != nil {
+		return errors.Annotate(req.err, "put into metadata failed")
+	}
+
+	return nil
+}
+
 func (a *Append) resolve(startTS int64) bool {
 	latestTS := atomic.LoadInt64(&a.latestTS)
 	if latestTS <= 0 {
 		return false
 	}
 
-	startSecond := oracle.ExtractPhysical(uint64(startTS)) / int64(time.Second/time.Millisecond)
-	maxSecond := oracle.ExtractPhysical(uint64(latestTS)) / int64(time.Second/time.Millisecond)
-
-	if maxSecond-startSecond <= maxTxnTimeoutSecond {
+	pbinlog, err := a.readBinlogByTS(startTS)
+	if err != nil {
+		log.Error(errors.ErrorStack(err))
 		return false
 	}
 
-	pbinlog, err := a.readBinlogByTS(startTS)
+	resp, err := a.helper.GetMvccByEncodedKey(pbinlog.PrewriteKey)
 	if err != nil {
-		log.Error(errors.ErrorStack(err))
+		log.Error("GetMvccByEncodedKey failed", zap.Error(err))
+	} else if resp.RegionError != nil {
+		log.Error("GetMvccByEncodedKey failed", zap.Stringer("RegionError", resp.RegionError))
+	} else if len(resp.Error) > 0 {
+		log.Error("GetMvccByEncodedKey failed", zap.String("Error", resp.Error))
+	} else {
+		for _, w := range resp.Info.Writes {
+			if !(int64(w.StartTs) == startTS) {
+				continue
+			}
+
+			if w.Type != kvrpcpb.Op_Rollback {
+				// Sanity checks
+				if int64(w.CommitTs) <= startTS {
+					log.Warn("op type not Rollback, but have unexpect commit ts",
+						zap.Int64("startTS", startTS),
+						zap.Uint64("commitTS", w.CommitTs))
+					return false
+				}
+
+				err := a.writeCBinlog(pbinlog, int64(w.CommitTs))
+				if err != nil {
+					log.Error("writeCBinlog failed", zap.Error(err))
+					return false
+				}
+			} else {
+				// Will get the same value as start ts if it's rollback, set to 0 for log
+				w.CommitTs = 0
```

Comment:
no need to change it

Reply:
easy to see whether it is commited (o rollback), and keep it same as the log of L471( the case call `GetTxnStatus`)